Simplify StructureID encoding scheme.

MenloDorian opened the pull request:
#### 6eac446be4449cb0c94d52b8ee43926a53bbb180
<pre>
Simplify StructureID encoding scheme.
<a href="https://bugs.webkit.org/show_bug.cgi?id=304017">https://bugs.webkit.org/show_bug.cgi?id=304017</a>
<a href="https://rdar.apple.com/166322205">rdar://166322205</a>

Reviewed by Dan Hecht and Yusuke Suzuki.

We can just use a base + offset encoding scheme for all CPU(ADDRESS64) platforms.  Here are the
details:

1. There&apos;s no need to mask the StructureID before adding it to the base address.  This was
   originally introduced as a caging security mitigation.

   However, it only worked if the Structure heap size if 4GB.   On iOS (and other embedded
   platforms), this caging was never done, and cannot be done because of the STRUCTURE_ID_WITH_SHIFT
   encoding.  The caging is also of very low value as a security mitigation.  Since it does not
   add much value, we&apos;ll remove it to simplify things.

2. Previously, for macOS, we encode StructureID as a 32-bit offset from the start of that 4G
   aligned region.

   Instead of doing this, we now encode StructureID as the bottom 32-bits of the 64-bit Structure
   address.  This is partially made possible by the removal of masking in (1), and it simplifies
   the code a lot.  To decode a StructureID, we simple add structureIDBase to the StructureID,
   where structureIDBase is startOfStructureHeap with the lower 32-bits masked to 0.

3. Previously, to enable the StructureID encoding scheme in (2), we always require that the
   structure heap start address be 4GB aligned.

   Now that StructureID is just the lower 32-bits of the address, the structure heap does not
   need to start on a 4GB region.  We only require that the upper 32-bits of the entire structure
   heap region be constant.  To achieve this, we require that the structure heap size be a power
   of 2 (which was always the case), and that the structure heap start address be aligned on the
   structure heap size.  With that we guarantee that the structure heap will never span across
   two 4GB granules i.e. the top 32-bits of the Structure address is guaranteed to be the same
   for any address in the structure heap.

4. Previously, iOS (and other embedded platforms), we needed to use the STRUCTURE_ID_WITH_SHIFT
   encoding for StructureID.  This is because on these platforms, it is not always possible to
   allocate the structure heap starting at a 4G boundary.  The number of available 4G boundaries
   are few.

   However, with (3), the structure heap only need to be aligned on the heap size.  Since it is
   reasonable for the heap size to be smaller on embedded platforms, the number of such aligned
   boundaries are significantly increases, and we should be able to successfully allocate the
   structure heap on the needed boundary.  Hence, we can now switch all embedded platforms over
   to using the same StructureID encoding as macOS.  The STRUCTURE_ID_WITH_SHIFT can now be
   completely removed.

5. This also allows us to shrink Structure&apos;s atomSize back to 16 bytes.  Currently, that doesn&apos;t
   make a difference, but it opens the door for potential size optimzations.

6. Add Options::structureHeapSizeInKB() to allow us to more easily test and experiment with
   different structure heap sizes.

7. Added more dump info to the RELEASE_ASSERTs in the StructureMemoryManager constructor.
   These will help us diagnose crashes, should we get crashes due to failed structure heap
   memory reservation.

This functionality should be covered by existing JSC and WebKit tests as Structures are used
prolifically.  This patch also adds more ASSERTs and RELEASE_ASSERTs to help catch issues
sooner in those tests.

* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/heap/StructureAlignedMemoryAllocator.cpp:
(JSC::StructureMemoryManager::StructureMemoryManager):
(JSC::StructureAlignedMemoryAllocator::initializeStructureAddressSpace):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::emitNonNullDecodeZeroExtendedStructureID):
(JSC::AssemblyHelpers::emitEncodeStructureID):
* Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/JSCConfig.h:
(JSC::structureIDBase):
* Source/JavaScriptCore/runtime/MegamorphicCache.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::Structure):
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureID.h:
(JSC::StructureID::decode const):
(JSC::StructureID::tryDecode const):
(JSC::StructureID::encode):
* Source/JavaScriptCore/tools/IntegrityInlines.h:
(JSC::Integrity::auditStructureID):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::decodeNonNullStructure):
(JSC::Wasm::OMGIRGenerator::encodeStructureID):

Canonical link: <a href="https://commits.webkit.org/304344@main">https://commits.webkit.org/304344@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87f7fc0787a8e5b8c8e38571c60fa831f149d898

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135315 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46583 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142818 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4077ae4b-9aaa-46f7-96ce-68548cadd553) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137184 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8334 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7543 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103404 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/564058b8-ae7c-4da2-84f4-cdf5f6e51e58) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5957 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121274 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/84268 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c1ee5437-9773-412d-ac04-ecf742aa4e8e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5741 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3343 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3404 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127310 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39456 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145513 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133799 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7379 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40026 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111781 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7421 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/6179 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112151 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5589 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117574 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20862 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7434 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35700 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/166648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7184 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70980 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/166648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7405 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7285 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/heap/StructureAlignedMemoryAllocator.cpp:
```diff
@@ -94,16 +95,28 @@ class StructureMemoryManager {
-        static_assert(hasOneBitSet(structureHeapAddressSize));
-        uintptr_t mappedHeapSize = structureHeapAddressSize;
+        uintptr_t preferredStructureHeapSize = structureHeapAddressSize;
+        if (Options::structureHeapSizeInKB())
+            preferredStructureHeapSize = static_cast<uintptr_t>(Options::structureHeapSizeInKB()) * KB;
```
How about setting the default value of `structureHeapSizeInKB` to `structureHeapAddressSize / KB`?

MenloDorian replied:
That will create a dependency of `OptionsList.h` on `StructureID.h`.  I think this is fine as is.  This also makes it clear whether the user specified an override or not when we dump options for inspection.

Commit:
Simplify StructureID encoding scheme.
https://bugs.webkit.org/show_bug.cgi?id=304017
rdar://166322205

Reviewed by Dan Hecht and Yusuke Suzuki.

We can just use a base + offset encoding scheme for all CPU(ADDRESS64) platforms.  Here are the
details:

1. There's no need to mask the StructureID before adding it to the base address.  This was
   originally introduced as a caging security mitigation.

   However, it only worked if the Structure heap size if 4GB.   On iOS (and other embedded
   platforms), this caging was never done, and cannot be done because of the STRUCTURE_ID_WITH_SHIFT
   encoding.  The caging is also of very low value as a security mitigation.  Since it does not
   add much value, we'll remove it to simplify things.

2. Previously, for macOS, we encode StructureID as a 32-bit offset from the start of that 4G
   aligned region.

   Instead of doing this, we now encode StructureID as the bottom 32-bits of the 64-bit Structure
   address.  This is partially made possible by the removal of masking in (1), and it simplifies
   the code a lot.  To decode a StructureID, we simple add structureIDBase to the StructureID,
   where structureIDBase is startOfStructureHeap with the lower 32-bits masked to 0.

3. Previously, to enable the StructureID encoding scheme in (2), we always require that the
   structure heap start address be 4GB aligned.

   Now that StructureID is just the lower 32-bits of the address, the structure heap does not
   need to start on a 4GB region.  We only require that the upper 32-bits of the entire structure
   heap region be constant.  To achieve this, we require that the structure heap size be a power
   of 2 (which was always the case), and that the structure heap start address be aligned on the
   structure heap size.  With that we guarantee that the structure heap will never span across
   two 4GB granules i.e. the top 32-bits of the Structure address is guaranteed to be the same
   for any address in the structure heap.

4. Previously, iOS (and other embedded platforms), we needed to use the STRUCTURE_ID_WITH_SHIFT
   encoding for StructureID.  This is because on these platforms, it is not always possible to
   allocate the structure heap starting at a 4G boundary.  The number of available 4G boundaries
   are few.

   However, with (3), the structure heap only need to be aligned on the heap size.  Since it is
   reasonable for the heap size to be smaller on embedded platforms, the number of such aligned
   boundaries are significantly increases, and we should be able to successfully allocate the
   structure heap on the needed boundary.  Hence, we can now switch all embedded platforms over
   to using the same StructureID encoding as macOS.  The STRUCTURE_ID_WITH_SHIFT can now be
   completely removed.

5. This also allows us to shrink Structure's atomSize back to 16 bytes.  Currently, that doesn't
   make a difference, but it opens the door for potential size optimzations.

6. Add Options::structureHeapSizeInKB() to allow us to more easily test and experiment with
   different structure heap sizes.

7. Added more dump info to the RELEASE_ASSERTs in the StructureMemoryManager constructor.
   These will help us diagnose crashes, should we get crashes due to failed structure heap
   memory reservation.

This functionality should be covered by existing JSC and WebKit tests as Structures are used
prolifically.  This patch also adds more ASSERTs and RELEASE_ASSERTs to help catch issues
sooner in those tests.

* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/heap/StructureAlignedMemoryAllocator.cpp:
(JSC::StructureMemoryManager::StructureMemoryManager):
(JSC::StructureAlignedMemoryAllocator::initializeStructureAddressSpace):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::emitNonNullDecodeZeroExtendedStructureID):
(JSC::AssemblyHelpers::emitEncodeStructureID):
* Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/JSCConfig.h:
(JSC::structureIDBase):
* Source/JavaScriptCore/runtime/MegamorphicCache.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::Structure):
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureID.h:
(JSC::StructureID::decode const):
(JSC::StructureID::tryDecode const):
(JSC::StructureID::encode):
* Source/JavaScriptCore/tools/IntegrityInlines.h:
(JSC::Integrity::auditStructureID):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::decodeNonNullStructure):
(JSC::Wasm::OMGIRGenerator::encodeStructureID):

Canonical link: https://commits.webkit.org/304344@main
